backend/src/Controllers/castVoteController.ts

```diff
@@ -0,0 +1,187 @@
+import { Election, electionValidation } from "../../../domain_model/Election";
+import { ElectionRoll, ElectionRollState } from "../../../domain_model/ElectionRoll";
+import { Ballot, Ballot as BallotType, ballotValidation } from '../../../domain_model/Ballot';
+import { IRequest } from "../IRequest";
+import ServiceLocator from "../ServiceLocator";
+import Logger from "../Services/Logging/Logger";
+import { BadRequest, InternalServerError, Unauthorized } from "@curveball/http-errors";
+import { ILoggingContext } from "../Services/Logging/ILogger";
+import {expectUserFromRequest, expectValidElectionFromRequest, catchAndRespondError} from "./controllerUtils";
+import { randomUUID } from "crypto";
+
+const ElectionsModel = ServiceLocator.electionsDb();
+const ElectionRollModel = ServiceLocator.electionRollDb();
+const BallotModel = ServiceLocator.ballotsDb();
+const CastVoteStore = ServiceLocator.castVoteStore();
+
+
+async function castVoteController(req: IRequest, res: any, next: any) {
+    Logger.info(req, "Cast Vote Controller");
+
+    const user = expectUserFromRequest(req);
+    const inputBallot:Ballot = req.body.ballot;
+    const validationErr = ballotValidation(inputBallot);
+    if (validationErr){
+        const errMsg = "Invalid Ballot: "+ validationErr
+        Logger.info(req, errMsg);
+        throw new BadRequest(errMsg);
+    }
+
+    //TODO: currently we have both a value on the input Ballot, and the route param.
+    //do we want to keep both?  enforce that they match?
+    const targetElectionId = req.params.id;
+    inputBallot.election_id = targetElectionId;
+    const targetElection = await ElectionsModel.getElectionByID(targetElectionId, req);
```

Comment:
Election data should already be available in req.election by this point from the param middleware in election.routes.js.

Probably wouldn't hurt to enforce that the ids match.

Reply:
There are a bunch of tests where the ids don't match, so will leave enforcing this as a TODO for another diff.